### ISSUE TITLE
Knowledge scope discriminator questions

### ISSUE THREAD

**Issue by 3514000142** [ecoxial2007/CH-ICL#5]
Hello, I'm very interested in your paper. I'm currently trying to reproduce the experiment, but I have some problems during the reproduction process.
(1)Following the process provided on github, I downloaded the dataset and other files from Baidu Netdisk and first ran them using trainval.sh (python3 src/trainval.py \ --dataset 'peir' \...) I trained this, and one round was over. Then I ran (python3 src/trainval.py \ --dataset 'pathvqa'), generating checkpoints for the peir file and checkpoints for the pathvqa file. (The checkpoints in my pathvqa file are only 70M, while yours is 800M. The same goes for radvqa and slake. I'm not sure if there's a problem here.) Then, change the checkpoints path in the testing section. I generated pathvqa_results.json and pathvqa_keyword_results.json, but I didn't find the Knowledge scope discriminator. The two files I generated, pathvqa_results.json and pathvqa_keyword_results.json, are almost exactly the same. Is the keyword "Knowledge scope discriminator"? When running the program, the keywords and candiates contents are exactly the same, as shown in the figure. How can I solve this? Is there a problem with the process of running the program?
(2) Where should I look for the metrics in the Text Generation Quality Assessment section? After I ran python3 tools/ openai_api_test_keyword-.py, only the overall accuracy rate was displayed in the end. Do I need to modify the code myself to display "open" and "close"?
(3) I ran the code using the local model, but the results were quite different. Maybe there was a problem with the code I modified. Could you provide the code using the local model?
I'm very sorry to disturb you. Looking forward to your reply!! thank you

<img width="1370" height="496" alt="Image" src="https://github.com/user-attachments/assets/3f439ec4-d583-48f6-b60b-82a28d0d6764" />[](url)

**Comment by ecoxial2007** [comment]
Thank you for your interest in our paper. Here are the answers to your questions:

1. The training process only saves the final feature fusion module. To get the 800M end-to-end checkpoint, you need to manually combine it with the biomedicalCLIP weights. The similar results you're seeing are expected because the PEIR and PathVQA datasets have some data overlap, and the In-Context Learning (ICL) method functions as an ensemble.

2. Yes, you need to modify the code manually to display those specific metrics.

3. I'm not entirely sure what you mean by a "local model."

- If you are using smaller models (e.g., 7B scale), instruction-tuned versions may not perform well. This task requires a large base model like GPT-4o/Gemini/Deepseek, as instruction tuning can impair in-context learning capabilities. You can try different base models by modifying the prompt.

- If you are referring to the fine-tuned fusion module, it requires careful and repeated hyperparameter tuning (learning rate, optimizer, batch size, epochs) to achieve the best results due to the small and long-tailed dataset. For PathVQA, this can sometimes mean manually stopping training, loading the best model, and resuming with a lower learning rate.

Hope this helps!

**Comment by 3514000142** [comment]
> Thank you for your interest in our paper. Here are the answers to your questions:
> 
> 1. The training process only saves the final feature fusion module. To get the 800M end-to-end checkpoint, you need to manually combine it with the biomedicalCLIP weights. The similar results you're seeing are expected because the PEIR and PathVQA datasets have some data overlap, and the In-Context Learning (ICL) method functions as an ensemble.
> 2. Yes, you need to modify the code manually to display those specific metrics.
> 3. I'm not entirely sure what you mean by a "local model."
> 
> * If you are using smaller models (e.g., 7B scale), instruction-tuned versions may not perform well. This task requires a large base model like GPT-4o/Gemini/Deepseek, as instruction tuning can impair in-context learning capabilities. You can try different base models by modifying the prompt.
> * If you are referring to the fine-tuned fusion module, it requires careful and repeated hyperparameter tuning (learning rate, optimizer, batch size, epochs) to achieve the best results due to the small and long-tailed dataset. For PathVQA, this can sometimes mean manually stopping training, loading the best model, and resuming with a lower learning rate.
> 
> Hope this helps!

Hello, thank you for your reply. After adding the biomedicalCLIP weights to the model_feature-.py file, I ran the trainval.py file to generate 800M checkpoints. To change the checkpoints path of the radvqa dataset, run the test.py file to generate radvqa_results.json. To change the peir dataset path, run the test_peir file to generate radvqa_keyword_results.json. The contents of these two json files are still exactly the same. The same goes for using your checkpoint (I uploaded the attachments of these two files, and the contents of the LLM run are almost exactly the same as those of these two json files). The knowledge scope and candidates parts in your thesis are completely different. It's this part that I ran out and it's different from your paper. Also, regarding the PMC-LLaMA-7B you are using, github only has code that can be run via api. Could you provide the code that can be run using PMC-LLaMA-7B?

If possible, may I add your contact information? It's 3514000142. This is my QQ.
Thank you.

<img width="1359" height="280" alt="Image" src="https://github.com/user-attachments/assets/c35ef6c7-a69c-45cc-b3a3-cdb923db8204" />

[radvqa_keyword_results.json](https://github.com/user-attachments/files/22114562/radvqa_keyword_results.json)
[radvqa_results.json](https://github.com/user-attachments/files/22114563/radvqa_results.json)

）

**Comment by ecoxial2007** [comment]
Hi,

Thanks for reaching out.

This project was built two years ago and is no longer actively maintained. I haven't run the code, but after a brief review, I found a small error in test_peir.py. The import should be from model_peir, not from model. I have corrected this in the repository, so please pull the latest version and try again.

Regarding the 7B model, our experimental results align with your observations. These smaller models tend to only answer questions with clear-cut correct candidates, like "yes" or "no". For this reason, we don't recommend using them.

If you must use a smaller model, you could try modifying the prompt to force a response, for example: "You must choose the closest answer from the candidates provided and are not allowed to refuse." You should also adjust the temperature and top_k parameters. However, please be aware that this will likely decrease the overall accuracy.

**Comment by 3514000142** [comment]
> Hi,
> 
> Thanks for reaching out.
> 
> This project was built two years ago and is no longer actively maintained. I haven't run the code, but after a brief review, I found a small error in test_peir.py. The import should be from model_peir, not from model. I have corrected this in the repository, so please pull the latest version and try again.
> 
> Regarding the 7B model, our experimental results align with your observations. These smaller models tend to only answer questions with clear-cut correct candidates, like "yes" or "no". For this reason, we don't recommend using them.
> 
> If you must use a smaller model, you could try modifying the prompt to force a response, for example: "You must choose the closest answer from the candidates provided and are not allowed to refuse." You should also adjust the temperature and top_k parameters. However, please be aware that this will likely decrease the overall accuracy.

您好，非常抱歉再次打扰您，对于keywords文件我将trainval_single_batch_get_keyword.py文件中的第32行的 with open(f'{args.dataset}_results.json', 'r') as f:改为 with open(f'{args.dataset}_keywords_results.json', 'r') as f:后我可以正常生成文件了，我不清楚是不是这里的问题，由于我计划将您的工作作为baseline，我还有一些问题想要请教：
1.对于trainval.py文件的198行if overall_acc > 0.801: quit()这句代码我保留会导致slake数据集的指标很低，不理解这句代码的含义
2.因为biomedclip的效果最好，所以您的百度云盘是只提供了biomedclip的文件吗？
3.我更改了学习率以及seed，但是使用您微调过的模型（不使用大模型）训练测试slake数据集和pathvqa数据集指标总是差了很多，使用您的检查点进行测试也达不到论文的水平，能告诉我接下来我该怎么做吗？
4.由于我自己编写代码去使用PMC-LLAMA-7B模型，但是我写的代码总是有问题，所以您的使用PMC-LLAMA-7B模型的运行的代码可以发我一份吗？
真的是非常抱歉再次打扰您，期待您的回复！！！

**Comment by ecoxial2007** [comment]
您好，我不确定您说的低具体是多少，也不确定您是如何进行测试的。如果直接使用test.sh，在使用我提供的正确权重、正确特征，代码不出错的情况下，理论上SLAKE应该有85.0%，PathVQA应该有61.8%。

但需要注意的是，这两个数据集的答案列表里都有同义，但严格匹配会判错的情况，例如“yellow”、"yellow color"，因此我们**利用字符串匹配判的准确率**，例如 pred = "yellow"， GT="yellow color" 也算正确，你也可以调用GPT-API（比如最便宜的4o）判定。

其他问题：
1. quit() 是debug时留下的，无意义
2. 是
3. 关于训练+调参的问题：三个数据集的数量很小，batch、head、lr、dropout一系列参数都会导致精度有+-10%的波动，请结合early stop 以及手动缩小学习率调参，看情况调整batch；你也可以对标签进行清洗和统一。
4. PMC-LLAMA-7B会因为拒绝回答导致准确率低于基线，不建议使用。如果非要使用小模型，也最好不要使用经过专业领域指令微调的模型，找基础能力更强的比如Qwen2.5会好一些。
